src/vision/status_quo/niklaus_lost_his_context.md

```diff
@@ -0,0 +1,99 @@
+
+fn load_image(src: &url, inside_of: Index<Node>, context: &mut Context<'_>) {
+    spawn_local(async move {
+        let url = src.to_string();
+        let request = //still elided
+        let data = window().unwrap().fetch_with_request(&request).await.unwrap().etc.etc.etc;
+        let image = parse_png(data, context);
+        context.display_tree.add_child_node(inside_of, image);
+    });
+}
+```
+
+Well, that stopped compiling. After looking through far too many compiler error messages, Niklaus realizes that the problem is the context itself. You see, whenever you close over some borrowed state, your closure's lifetime is bounded by that borrow. This is reasonable for closures, as there are plenty of contexts where you would pass such a closure and have it called immediately. However, the whole point of futures is that they are *not* called immediately - you hold them inside of a thing called an "executor", which typically needs futures that own their contents.
+
+At this point, Niklaus is looking at either rewriting a good chunk of the program to not use contexts, or rewriting his own code to sidestep the problem. He chooses the latter, by holding a reference to the thing that produces the context and then updating it after the async portion of the task has concluded:
+
+```use crate::{Index, Player}
+use wasm_bindgen_futures::spawn_local;
+use web_sys::window;
+use crate::parse_png;
+
+fn load_image(src: &url, inside_of: Index<Node>, player: Arc<Mutex<Player>>) {
+    spawn_local(async move {
+        let url = src.to_string();
+        let request = //still elided
+        let data = window().unwrap().fetch_with_request(&request).await.unwrap().etc.etc.etc;
+
+        player.lock().unwrap().update(|context| {
+            let image = parse_png(data, context);
+            context.display_tree.add_child_node(inside_of, image);
+        });
+    });
+}
+```
+
+It works, but it causes problems. Async and contextual code can't really coexist; so async code always has to live at the periphery of the program. More complicated async behaviors, such as reading the image in chunks and progressively decoding it, requires a delicate dance of await followed by update and so on. Niklaus tells Alan about this and he says it has something to do with colored functions, something he isn't quite familiar about.
+
+## 🤔 Frequently Asked Questions
+
+* **What are the morals of the story?**
+    * Borrowing means that Rust has three function colors, not two:
+      * Sync, which can both own and borrow it's parameters
+      * Owned Async, which can only own parameters
+      * Borrowed Async, which can both own and borrow parameters, but gains a lifetime
+    * Non-`'static` Futures are of limited use as lifetimes are tied to the sync stack. I think you can use `block_on` with them but that's about it.
+    * Borrowed Async functions would be useful if there was a way to have a future whose borrows were scoped to the moment they are `poll`ed.
```

Comment:
These seem to be the real crux of the issue. This makes it clearer to me that seeing the error messages would really drive this point home. Perhaps it would be good to show more of the struggle in the time between getting the initial error message and figuring out a solution.